### PR TITLE
[IMP] crm: add menuitem "My Activities" and new view

### PR DESCRIPTION
**Pull request by pierrecollinet**
Description of the issue/feature this PR addresses:
0) Remove the decoration-danger set on the CRM tree views 
Purpose : We'll set it on the right fields instead so that it's clearer for the user.

Add activity_ids with list_activity widget here : https://tinyurl.com/ycqncun2 
And the remaining_days on this field here :https://tinyurl.com/y73f32t5 
    
1) Separate Leads and Opportunities in the systray. There should be one line for each type
NB : It's been done for mailings in case you're looking for an example https://tinyurl.com/ybcms689

2) Add a new menu item "My Activities" https://docs.google.com/presentation/d/1EDZx9rdSTkmgFRXDitp45OXRL0EI3PJcUMvjmFLhVgU/edit#slide=id.g74c54ce7d7_0_0
Domain of this action is : crm.lead, type = opportunity at least an activity set
Filter is : My Activities (at least an activity assigned to me)

3) Create this new view for opportunities
https://docs.google.com/presentation/d/1EDZx9rdSTkmgFRXDitp45OXRL0EI3PJcUMvjmFLhVgU/edit#slide=id.p

And this one for leads :
https://docs.google.com/presentation/d/1EDZx9rdSTkmgFRXDitp45OXRL0EI3PJcUMvjmFLhVgU/edit#slide=id.g773783b040_0_4

4) The path from the systray (opportunity/leads) done in 1) should redirect to the respective new view instead of on the normal pipeline.
On top of the existing filters/domain, it adds the time-based filter : Today, Late, Future Activities, ...

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/crm/models/crm_lead.py [hunk 1]
```diff
@@ -764,6 +761,53 @@ def action_schedule_meeting(self):
...
+            'binding_model': "crm.lead",
+            'binding_views': "list",
+            'id': "crm.action_lead_mass_mail",
+            'target': 'new',
+            'context': ctx,
+        }
+
+    def action_send_sms(self):
+        self.ensure_one()
+        if not self.phone:
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:787]: Actually lead has both phone and mobile + sanitized versions of it available. I think this condition could be improved accordingly :) .

File: addons/crm/models/crm_lead.py [hunk 2]
```diff
@@ -764,6 +761,53 @@ def action_schedule_meeting(self):
...
+            raise UserError(_("There is no professional mobile for this employee."))
+        context = dict(self.env.context)
+        context.update(default_composition_mode="comment", default_mass_keep_log=True, default_res_id=self.id, default_res_model="crm.lead")
+        return {
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:791]: Well this should be located in crm_sms module as crm doe not depend on SMS. SMS is auto install but adds bridge modules. It is not directly integreated in crm.

File: addons/crm/models/crm_lead.py [hunk 3]
```diff
@@ -764,6 +761,53 @@ def action_schedule_meeting(self):
...
+            'view_mode': 'form',
+            'res_model': 'mail.compose.message',
+            'binding_model': "crm.lead",
+            'binding_views': "list",
+            'id': "crm.action_lead_mass_mail",
+            'target': 'new',
+            'context': ctx,
+        }
+
+    def action_send_sms(self):
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:785]: Can't you read actions already defined in crm_sms like crm_lead_act_window_sms_composer_multi or crm_lead_act_window_sms_composer_single (funny, names are the inverse of what they should do) and simply update some missing pieces ? That way we avoid doing a whole new action by hand.

File: addons/crm/models/crm_lead.py [hunk 4]
```diff
@@ -764,6 +761,53 @@ def action_schedule_meeting(self):
         }
         return action
 
+    def action_send_mail(self):
+        self.ensure_one()
+        template = self.env.ref('crm.email_template_opportunity_mail')
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:766]: Can't you re-use action_lead_mass_mailand tweak it, like explained for sms sending ? Seems easier.

File: addons/crm/models/crm_lead.py [hunk 5]
```diff
@@ -483,14 +485,11 @@ def write(self, vals):
     @api.model
     def search(self, args, offset=0, limit=None, order=None, count=False):
         """ Override to support ordering on activity_date_deadline_my.
-
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:486]: Linting: do not remove blank lines ... this makes unnecessary diff and those blank lines help reading ^^ .

File: addons/crm/models/crm_lead.py [hunk 6]
```diff
@@ -10,6 +10,7 @@
 from odoo.osv import expression
 from odoo.tools.translate import _
 from odoo.tools import email_re, email_split
+from odoo.tools.date_utils import add
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:13]: Is it necessary to use add although a simple standard timedelta / relativedelta could do it ?

File: addons/crm/models/crm_lead.py [hunk 7]
```diff
@@ -764,6 +761,53 @@ def action_schedule_meeting(self):
...
+            "view_type": 'form',
+            "type": 'ir.actions.act_window',
+            "target": "new",
+            "context": context,
+        }
+
+    def action_snooze(self):
+        self.ensure_one()
+        my_next_activity = self.activity_ids.filtered(lambda activity: activity.user_id == self.env.user)[:1]
+        my_next_activity.write({
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:806]: Pay attention you don't necessarily have an activity.

File: addons/crm/static/src/js/systray_activity_menu.js [hunk 8]
```diff
@@ -2,25 +2,64 @@ odoo.define('crm.systray.ActivityMenu', function (require) {
...
+     * @private
+     * @param {MouseEvent} event
+     */
+    _onActivityFilterClick: function (event) {
+        // fetch the data from the button otherwise fetch the ones from the parent (.o_mail_preview).
+        var action;
+        var data = _.extend({}, $(event.currentTarget).data(), $(event.target).data());
+        var additional_context = {};
+
+        if (data.model_name === 'Lead'){
```
Comment (tde-banana-odoo) [inline addons/crm/static/src/js/systray_activity_menu.js:35]: Not sure where model_name comes from, but it seems like it is some kind of string based on server data, that could be translated ... Shouldn't you propagate a type (lead_type) ?

File: addons/crm/static/src/js/systray_activity_menu.js [hunk 9]
```diff
@@ -2,25 +2,64 @@ odoo.define('crm.systray.ActivityMenu', function (require) {
...
+        } else if (data.model_name === 'Opportunity'){
+            action = 'crm.action_my_activities'
+        }
+
+        var domain = [['activity_ids.user_id', '=', session.uid]]
+        if (data.domain) {
+            domain = domain.concat(data.domain)
+        }
+
+        if (data.res_model === "crm.lead") {
```
Comment (tde-banana-odoo) [inline addons/crm/static/src/js/systray_activity_menu.js:46]: Coudln't we simplify by saying that
1/ check model -> it not lead, call super and return
2/ then do our internal popotte ?

File: addons/crm/views/crm_lead_views.xml [hunk 10]
```diff
@@ -332,7 +332,58 @@
                     <field name="message_needaction" invisible="1"/>
                     <field name="tag_ids" optional="hide" widget="many2many_tags" options="{'color_field': 'color'}"/>
                     <field name="priority" optional="hide"/>
-                    <field name="activity_exception_decoration" widget="activity_exception"/>
+                </tree>
+            </field>
+        </record>
+
+        <record id="crm_case_tree_view_my_activities" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:339]: Not sure why there are so much new views... I feel like only a new tree view is necessary. It could be based on existing view, using inherit_id (inherit from the based one) + mode = primary (becomes independent).

Inherit + primary allows to somehow "duplicate" a view: it became an independent view, still inheriting from the base one. It means that fields added to the base one are added to the inheriting one. You can then tweak your custom one with xpath to hide or remove unnecessary fields / add relevant one.

Also don't forget to give an higher priority for this view. Purpose is to avoid having it displayed by default (lower priority means default).

You can check for mode="primary" in the code to see examples :)

File: addons/crm/views/crm_lead_views.xml [hunk 11]
```diff
@@ -809,6 +870,44 @@ if record:
             <field name="action" ref="crm.crm_lead_all_leads"/>
         </record>
 
+        <!-- My Activities Menu -->
+        <record id="action_my_activities" model="ir.actions.act_window">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:874]: Quick guidelining: if you read guidelines, this should have an id like crm_lead_action_my_activities . See guidelines for more details :) normally there are examples in it.

File: addons/crm/views/crm_lead_views.xml [hunk 12]
```diff
@@ -809,6 +870,44 @@ if record:
...
+            <field name="view_mode">tree,kanban,graph,pivot,calendar,form,activity</field>
+            <field name="view_id" ref="crm_case_tree_view_leads2"/>
+            <field name="domain">[('type','=','lead')]</field>
+            <field name="search_view_id" ref="crm.view_crm_case_my_activities_filter"/>
+            <field name="context">{'default_type': 'lead',
+                    'search_default_assigned_to_me': 1}
+            </field>
+        </record>
+
+        <record id="crm_lead_my_activities" model="ir.ui.menu">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:900]: Another exemple of guidelining: crm_lead_menu_my_activities

File: addons/mail/static/src/js/systray/systray_activity_menu.js [hunk 13]
```diff
@@ -161,7 +161,7 @@ var ActivityMenu = Widget.extend({
         if (data.domain) {
             domain = domain.concat(data.domain)
         }
-        
+
```
Comment (tde-banana-odoo) [inline addons/mail/static/src/js/systray/systray_activity_menu.js:164]: Unnecessary diff here

File: addons/web/static/src/js/views/list/list_renderer.js [hunk 14]
```diff
@@ -857,6 +857,9 @@ var ListRenderer = BasicRenderer.extend({
      */
     _renderHeaderCell: function (node) {
         const { icon, name, string } = node.attrs;
+        if (node.attrs.name === 'activity_date_deadline_my') {
```
Comment (tde-banana-odoo) [inline addons/web/static/src/js/views/list/list_renderer.js:860]: Oopsie :)

Comment (tde-banana-odoo) [reply addons/web/static/src/js/views/list/list_renderer.js:860]: Still to remove

File: addons/crm/views/crm_lead_views.xml [hunk 15]
```diff
@@ -731,7 +716,108 @@
...
+                id="crm.action_lead_single_mail"
+                context="{
+                            'default_composition_mode': 'comment',
+                            'default_email_to':'{($object.email or \'\')|safe}',
+                            'default_use_template': True,
+                            'default_template_id': ref('crm.email_template_opportunity_mail'),
+                        }"
+        />
+
+        <record id="crm_case_tree_view_oppor" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:736]: In order to lessen the diff you should keep this one were it was defined first (l 583), otherwise we are a bit lost with the changes

File: addons/crm/views/crm_lead_views.xml [hunk 16]
```diff
@@ -731,7 +716,108 @@
...
-                />
+        />
+
+        <act_window name="Send email"
+                res_model="mail.compose.message"
+                binding_model="crm.lead"
+                binding_views="list"
+                view_mode="form"
+                target="new"
+                id="crm.action_lead_single_mail"
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:727]: Rename id to something like crm_lead_act_window_compose in order to follow guidelines ?

File: addons/crm/views/crm_lead_views.xml [hunk 17]
```diff
@@ -810,6 +896,40 @@ if record:
...
+            <field name="view_id" ref="crm_case_tree_view_my_activities"/>
+            <field name="domain">[('type','=','opportunity'), ('activity_ids','!=',False)]</field>
+            <field name="search_view_id" ref="crm.view_crm_case_my_activities_filter"/>
+            <field name="context">{'default_type': 'opportunity',
+                    'search_default_assigned_to_me': 1}
+            </field>
+        </record>
+
+        <!-- My Leads Action -->
+        <record model="ir.actions.act_window" id="action_my_leads2">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:913]: What is this "my leads 2 " ?

File: addons/crm/views/crm_lead_views.xml [hunk 18]
```diff
@@ -731,7 +716,108 @@
...
+                    <field name="probability" optional="hide"/>
+                    <field name="company_id" groups="base.group_multi_company" optional="show"/>
+                    <field name="tag_ids" optional="hide" widget="many2many_tags" options="{'color_field': 'color'}"/>
+                    <field name="referred" invisible="1"/>
+                    <field name="message_needaction" invisible="1"/>
+                </tree>
+            </field>
+        </record>
+
+        <record id="crm_case_tree_view_leads2" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:774]: Not sure it is used anywhere except in action myleads2 that does not seem to be used anywhere.

File: addons/crm/views/crm_lead_views.xml [hunk 19]
```diff
@@ -731,7 +716,108 @@
...
+                    />
+                </field>
+            </field>
+        </record>
+        <record id="crm_case_tree_view_my_activities" model="ir.ui.view">
+            <field name="name">crm.lead.tree.lead</field>
+            <field name="model">crm.lead</field>
+            <field name="mode">primary</field>
+            <field name="priority">15</field>
+            <field name="inherit_id" ref="crm.crm_case_tree_view_oppor"/>
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:800]: To move just below the crm_case_tree_view_oppor definition to ease reading

File: addons/crm/views/crm_lead_views.xml [hunk 20]
```diff
@@ -731,7 +716,108 @@
...
+                    <button name="%(crm.action_lead_single_mail)d" type="action" string="Email" icon="fa-envelope"
+                            context="{
+                                        'default_composition_mode': 'comment',
+                                        'default_use_template': True,
+                                    }"
+                    />
+                </field>
+            </field>
+        </record>
+        <record id="crm_case_tree_view_my_activities" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:795]: Right xml ID is crm_lead_view_list_activities (model + view (telling it is a view record) + list (as in list / form / kanban / ...) + suffix to help (see guidelines for more details).

File: addons/crm/views/crm_lead_views.xml [hunk 21]
```diff
@@ -731,7 +716,108 @@
...
+            <field name="name">crm.lead.tree.lead</field>
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:796]: Naming should be based on xml ID (replacing _ by ., see guidelines).

File: addons/crm_sms/views/crm_lead_views.xml [hunk 22]
```diff
@@ -26,4 +26,26 @@
             'default_res_id': active_id,
         }"
     />
-</odoo>
\ No newline at end of file
+
+    <record id="crm_case_tree_view_my_activities_extend" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/crm_sms/views/crm_lead_views.xml:30]: Inherited view should be named like an original one, as the module will be added as prefix for the full xml id.

Example: id="crm_lead_view_list_activities", full xml id is crm_sms.crm_lead_view_list_activities -> we know this is an inherit. Therefore to keep xml ids "short" we just use the same guidelines for naming inherited xml ids.

File: addons/crm_sms/views/crm_lead_views.xml [hunk 23]
```diff
@@ -26,4 +26,26 @@
...
+        <field name="name">crm.lead.tree.sms.button</field>
```
Comment (tde-banana-odoo) [inline addons/crm_sms/views/crm_lead_views.xml:31]: BUT name contains more details, it could be crm.lead.view.list.activities.inherit.sms

File: addons/website_crm_partner_assign/views/crm_lead_views.xml [hunk 24]
```diff
@@ -37,13 +37,24 @@
...
             <field name="arch" type="xml">
-                <field name="user_id" position="before">
+                <field name="priority" position="after">
                     <field name="partner_assigned_id" optional="show"/>
                     <field name="date_partner_assign" invisible="1"/>
                  </field>
             </field>
         </record>
 
+        <record id="view_crm_lead_geo_assign_tree" model="ir.ui.view">
```
Comment (tde-banana-odoo) [inline addons/website_crm_partner_assign/views/crm_lead_views.xml:47]: Same comments applies here for naming (you can go on the whole diff to see if everything is matching guidelines)

File: addons/crm/views/crm_lead_views.xml [hunk 25]
```diff
@@ -576,47 +576,114 @@
             </field>
         </record>
 
+        <act_window name="Send email"
+                res_model="mail.compose.message"
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:580]: HumanLinter: you have indentation issues (indentation should always be 4 spaces) -> be sure to configure your linter accordingly

File: addons/crm/views/crm_lead_views.xml [hunk 26]
```diff
@@ -576,47 +576,114 @@
...
+                binding_model="crm.lead"
+                binding_views="list"
+                view_mode="form"
+                target="new"
+                id="crm_lead_act_window_compose"
+                context="{
+                            'default_composition_mode': 'comment',
+                            'default_email_to':'{($object.email or \'\')|safe}',
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:588]: Normally setting a default email_to is not necessary as in comment mode actually you post on the record + email_template_opportunity_mail already defines email_to / partner_to

File: addons/crm/views/crm_lead_views.xml [hunk 27]
```diff
@@ -576,47 +576,114 @@
...
             crm.lead (as Opportunity) views
         -->
 
         <record id="crm_case_tree_view_oppor" model="ir.ui.view">
             <field name="name">crm.lead.tree.opportunity</field>
             <field name="model">crm.lead</field>
+            <field name="priority">20</field>
             <field name="arch" type="xml">
-                <tree string="Opportunities" decoration-muted="probability == 100" decoration-danger="activity_date_deadline and (activity_date_deadline &lt; current_date)">
+                <tree string="Opportunities" default_order="activity_date_deadline_my">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:603]: Meh, really not sure the standard list view should have an order on activity_date_deadline_my as this is linked to an hack. I am not sure it will work, and I think we should limit the grouping on activity_date_deadline_my only to "my activities" menu item. To validate with LNA but strongly pushing to limit the use of activity_date_deadline_my to your specific action.

File: addons/crm/views/crm_lead_views.xml [hunk 28]
```diff
@@ -576,47 +576,114 @@
...
+                <field name="user_id" position="attributes">
+                    <attribute name="optional">hide</attribute>
+                </field>
+                <field name="team_id" position="attributes">
+                    <attribute name="optional">hide</attribute>
+                </field>
+                <field name="message_needaction" position="after">
+                    <button name="action_snooze" type="object" string="Snooze 7d" icon="fa-bell-slash" />
+                    <button name="%(crm.crm_lead_act_window_compose)d" type="action" string="Email" icon="fa-envelope"
+                            context="{
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:652]: This context is already set on the action itself, it is required to set it again ?

File: addons/crm/views/crm_lead_views.xml [hunk 29]
```diff
@@ -576,47 +576,114 @@
...
+                </field>
+            </field>
+        </record>
+
+        <record id="view_crm_case_my_activities_filter" model="ir.ui.view">
+            <field name="name">crm.lead.search.myactivities</field>
+            <field name="model">crm.lead</field>
+            <field name="inherit_id" ref="crm.view_crm_case_leads_filter"/>
+            <field name="arch" type="xml">
+                <xpath expr="//filter[@name='activities_overdue']" position="replace">
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:666]: Can't you do something like
```
<xpath expr="//filter[@name='activities_overdue']" position="attributes">
    <attribute name="invisible">0</attribute>
</xpath>
```
This should allow to avoir redefining the whole filter.

Comment (pierrecollinet) [reply addons/crm/views/crm_lead_views.xml:666]: I tried, but it wasn't working... did I miss something?

Comment (tde-banana-odoo) [reply addons/crm/views/crm_lead_views.xml:666]: Well I am quite sure it worked, for example in mass_mailing_sms we use it to hide mail-related fields of mailing. What was the issue ? Not displaying, or view error ?

File: addons/crm/views/crm_lead_views.xml [hunk 30]
```diff
@@ -576,47 +576,111 @@
...
+                    <field name="message_needaction" invisible="1"/>
                 </tree>
             </field>
         </record>
 
+        <record id="crm_lead_view_list_activities" model="ir.ui.view">
+            <field name="name">crm.lead.list.activities</field>
+            <field name="model">crm.lead</field>
+            <field name="mode">primary</field>
+            <field name="priority">15</field>
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:639]: Priority 15 means this will be the default view.  I think this should be the inverse: setting 1 to opportunity view, 10 to lead view (as leads is an optional step that should not be the default view), and 20 to activity view, something like that.

File: addons/crm/views/crm_lead_views.xml [hunk 31]
```diff
@@ -576,47 +576,111 @@
...
+            <field name="model">crm.lead</field>
+            <field name="inherit_id" ref="crm.view_crm_case_leads_filter"/>
+            <field name="arch" type="xml">
+                <xpath expr="//filter[@name='activities_overdue']" position="replace">
+                    <filter string="Late Activities" name="activities_overdue"
+                            domain="[('activity_ids.date_deadline', '&lt;', context_today().strftime('%Y-%m-%d'))]"
+                            help="Show all opportunities for which the next action date is before today"/>
+                </xpath>
+                <xpath expr="//filter[@name='activities_today']" position="replace">
+                    <filter  string="Today Activities" name="activities_today"
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:669]: HumanLinter: extra spaces :D

File: addons/crm/views/crm_lead_views.xml (hunk 29 above)
Comment (tde-banana-odoo) [reply addons/crm/views/crm_lead_views.xml:663]: Up this one, if it does not work could you check with JS team maybe ?

File: addons/crm/views/crm_lead_views.xml [hunk 32]
```diff
@@ -810,6 +874,26 @@ if record:
...
+            <field name="view_id" ref="crm_lead_view_list_activities"/>
+            <field name="domain">[('type','=','opportunity'), ('activity_ids','!=',False)]</field>
+            <field name="search_view_id" ref="crm.view_crm_case_my_activities_filter"/>
+            <field name="context">{'default_type': 'opportunity',
+                    'search_default_assigned_to_me': 1}
+            </field>
+        </record>
+
+        <record id="crm_lead_action_my_activities_view_tree" model="ir.actions.act_window.view">
+            <field name="sequence" eval="1"/>
```
Comment (tde-banana-odoo) [inline addons/crm/views/crm_lead_views.xml:891]: <field name="sequence">1</field>

File: addons/crm/views/crm_lead_views.xml (hunk 29 above)
Comment (pierrecollinet) [reply addons/crm/views/crm_lead_views.xml:663]: I tried exactly that way but it still doesn't work. LNA tried also and got the same result.

I asked AAB and he told me his team has no time for the moment and it's nothing related to js. I will look at it more carefully.

Comment (tde-banana-odoo) [reply addons/crm/views/crm_lead_views.xml]: Ok. Then keep it like it currently, just add a comment above the redefinition telling why we do it like it and that this is temporary. That way we can move forward with this task and look at the underlying bug in master without being in a rush.

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@robodoo r+
